client/src/components/useFileUpload.ts

```diff
@@ -0,0 +1,202 @@
+    // When a file input dialog closes, it triggers a window focus event,
+    // which causes a refetch by default, so we turn that off.
+    // https://github.com/tannerlinsley/react-query/issues/2960
+    refetchOnWindowFocus: false,
+    refetchInterval: fileInfo => (isProcessing(fileInfo) ? 1000 : false),
+    // Once a file is finished processing or is deleted, call onFileChange
+    // (but don't call it the very first time we load the file)
+    onSuccess: fileInfo => {
+      if (isFirstFetch.current) {
+        isFirstFetch.current = false
+        return
+      }
+      if (!isProcessing(fileInfo) && options.onFileChange) {
+        options.onFileChange()
+      }
+    },
+    enabled: options.enabled,
+  })
+}
+
+export const useUploadFiles = (key: string[], url: string) => {
+  const [progress, setProgress] = useState<number>()
+
+  const putFiles = async (formData: FormData) => {
+    try {
+      await axios(url, addCSRFToken({
+        method: 'PUT',
+        data: formData,
+        onUploadProgress: progressEvent =>
+          setProgress(progressEvent.loaded / progressEvent.total),
+      }) as AxiosRequestConfig)
+    } catch (error) {
+      const { errors } = error.response.data
+      const message =
+        errors && errors.length ? errors[0].message : error.response.statusText
+      throw new ApiError(message, error.response.status)
+    } finally {
+      setProgress(undefined)
+    }
+  }
+
+  const queryClient = useQueryClient()
+
+  return {
+    ...useMutation<void, ApiError, FormData>(putFiles, {
+      onSuccess: () => queryClient.invalidateQueries(key),
+    }),
+    progress,
+  }
+}
+
+export const useDeleteFile = (key: string[], url: string) => {
+  const deleteFile = () => fetchApi(url, { method: 'DELETE' })
+  const queryClient = useQueryClient()
+  return useMutation<void, ApiError, void>(deleteFile, {
+    onSuccess: () => queryClient.invalidateQueries(key),
+  })
+}
+
+export const useBallotManifest = (
+  electionId: string,
+  jurisdictionId: string
+): IFileUpload => {
+  const url = `/api/election/${electionId}/jurisdiction/${jurisdictionId}/ballot-manifest`
+  const key = ['jurisdictions', jurisdictionId, 'ballot-manifest']
+  const uploadFiles = useUploadFiles(key, url)
+  const deleteFile = useDeleteFile(key, url)
+  const queryClient = useQueryClient()
+  return {
+    uploadedFile: useUploadedFile(key, url, {
+      onFileChange: () => {
+        // When the manifest changes, batch tallies and cvrs are reprocessed
+        queryClient.invalidateQueries([
+          'jurisdictions',
+          jurisdictionId,
+          'batch-tallies',
+        ])
+        queryClient.invalidateQueries(['jurisdictions', jurisdictionId, 'cvrs'])
+        // We need to reload the jurisdictions list with new file upload status
+        queryClient.invalidateQueries([
+          'elections',
+          electionId,
+          'jurisdictions',
+        ])
+      },
```

Comment:
Related to the above comment, why is it that we only need to invalidate queries on subsequent file upload and not initial file upload? Is it because the data we invalidate doesn't exist until the first file is uploaded?

Reply:
Yeah, this is the really tricky part of this whole thing. There are two things going on here:
1. When the page initially loads, there's a list of jurisdictions loaded (including file upload status for each jurisdiction). The modal used to be populated with data from that request. However, I didn't want to poll that endpoint during file uploads to get the status for a single file, since it can be pretty expensive (though maybe this was premature optimization...). It also seemed more straightforward for the file upload component+hook to be self-contained and only think about one file. So I needed some way to reload the jurisdiction list after a file upload completed. Since file uploads are only "complete" after background processing is finished, it's not as simple as just reloading once the initial upload request completes - we need to wait until processing finishes. So that's why we invalidate the jurisdictions request on file _load_, but not on the first load (since there's no chance a file has changed on the first load).
2. In some audit types, there are two jurisdiction files to be uploaded (e.g. manifest and CVRs for ballot comparison). CVRs depend on manifests, since we need to know the batch names from the manifest in order to validate the CVR. That means if a user changes the manifest, the CVR needs to be reprocessed. We made a product decision a while back to do that automatically on manifest changes, which is nice from a user perspective and tricky technically. When a manifest file finishes processing, the server will automatically start the reprocessing of the existing CVR file. That means the frontend needs to know to start polling for the CVR file status after the manifest finishes processing. This is the other reason we use the `onFileChange` handler to invalidate CVRs (and batch tallies) on manifest file load, but not on the first load.

Confusing stuff! I wish there was a way to make it simpler, and I'm open to any suggestions.

Reply:
Wow, thank you for all this context!

I don't have any immediate thoughts on how to best simplify the larger architecture, but I do still think that it's a bit unexpected from an interface perspective for `onFileChange` to only run on subsequent file changes and not the first.

One alternative may be to call the option `onUploadSuccess` and expose a param indicating whether it's the initial upload, which consumers can then use

```ts
onUploadSuccess: (isInitialUpload?: boolean) => void
```

Reply:
I meant `onFileChange` to mean "when a file is either successfully uploaded, uploaded but fails processing, or is deleted" - in other words, whenever the state of the uploaded file in the database changes. That's why it doesn't run on first load, because the uploaded file didn't change. I'm pretty confident that's the behavior we want. Would making that more clear (through name or comment) help?

Reply:
Ah I see now! The source of my confusion was a misinterpretation of the phrase "first load." I was interpreting it as the first upload when it's actually referring to loading of a previously uploaded file (if any)

Tweaking the existing explanatory comment could be helpful.

If I'm still missing something (very possible haha), happy to hop on a quick chat!